### PR TITLE
Made Upgradeable version

### PR DESCRIPTION
**Pull request by MrMcGoats**
Made an upgradeable version of ERC721A

Also added hardhat files, so the contract can easily be built and deployed.

Tests by @cygaar (https://github.com/cygaar/ERC721A/tree/version2)

### CODE REVIEW

**Comment by AndreiD** [comment]
dude, you have your private key in .env-example...

**Comment by MrMcGoats** [comment]
That's not my private key; that's the publicly known private key that hardhat uses for tests

**Comment by cygaar** [comment]
Thanks for putting this together - as you know we've been working on shaping the repo, which will be done in the next couple days. Will review this PR soon

**Comment by MrMcGoats** [comment]
I also made a version of the upgradeable contract that extends OpenZeppelin's ERC721Upgradeable. I needed ERC721A to have the legacy _mint function for the project I'm working on (Xenum) and that seemed to be the simplest way.

Wasn't sure if that's something you want in this repo or not. Should I merge that change into this PR too?

**Comment by AndreiD** [comment]
@MrMcGoats I was looking at this too. Can you name it ERC721AUpgradable.sol and put it in the same place maybe?

**Comment by cygaar** [comment]
Can you add the legacy _mint function in a separate PR? I'd like to keep different features in different PRs so we can do a better job reviewing each one

**Comment by MrMcGoats** [comment]
Oh, sure. Sorry, thought @AndreiD was asking me to merge it into here. I'll revert those merges.

I'll add it to the non-upgradeable version too and submit a new PR

File: scripts/deploy.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  // Hardhat always runs the compile task when running scripts with its command
+  // line interface.
+  //
+  // If this script is run directly using `node` you may want to call compile
+  // manually to make sure everything is compiled
+  // await hre.run('compile');
+
+  // We get the contract to deploy
+  const Contract = await ethers.getContractFactory("ERC721A");
+  const contract = await Contract.deploy();
```
Comment (coindegen) [inline scripts/deploy.js:18]: Maybe I'm missing something, but don't upgradeable contracts use  `await upgrades.deployProxy(Contract)`? Reference: https://docs.openzeppelin.com/contracts/4.x/upgradeable#usage

Comment (AndreiD) [reply scripts/deploy.js:18]: here's what I use, maybe it helps
```
      const instanceContract = await upgrades.deployProxy(CONTRACT, {
		kind: "uups",
		log: true,
		skipIfAlreadyDeployed: true,
	})
	
	with ugprading at
	
		// We get the contract to deploy
	const proxyAddress = "0x............................."
	const contract = await ethers.getContractFactory("CONTRACTV2")
	console.log("Upgrading contract...")
	const upgr = await upgrades.upgradeProxy(proxyAddress, contract)
	await upgr.deployed()
	console.log("Contract upgraded @ ", upgr.address)

	const contractImplementation = await getImplementationAddress(upgr.address)
	console.log("++++++++++++++++++++++++++++++++++++++++++++++++")
	console.log("Contract Implementation:", contractImplementation)
	console.log("++++++++++++++++++++++++++++++++++++++++++++++++")
	
	await sleep(20000)
	await hre.run("verify:verify", {
		address: contractImplementation,
		constructorArguments: [],
	})
```

Comment (jplew) [reply scripts/deploy.js:18]: > here's what I use, maybe it helps
> 
> ```
> 
>       const instanceContract = await upgrades.deployProxy(CONTRACT, {
> 
> 		kind: "uups",
> 
> 		log: true,
> 
> 		skipIfAlreadyDeployed: true,
> 
> 	})
> 
> 	
> 
> 	with ugprading at
> 
> 	
> 
> 		// We get the contract to deploy
> 
> 	const proxyAddress = "0x............................."
> 
> 	const contract = await ethers.getContractFactory("CONTRACTV2")
> 
> 	console.log("Upgrading contract...")
> 
> 	const upgr = await upgrades.upgradeProxy(proxyAddress, contract)
> 
> 	await upgr.deployed()
> 
> 	console.log("Contract upgraded @ ", upgr.address)
> 
> 
> 
> 	const contractImplementation = await getImplementationAddress(upgr.address)
> 
> 	console.log("++++++++++++++++++++++++++++++++++++++++++++++++")
> 
> 	console.log("Contract Implementation:", contractImplementation)
> 
> 	console.log("++++++++++++++++++++++++++++++++++++++++++++++++")
> 
> 	
> 
> 	await sleep(20000)
> 
> 	await hre.run("verify:verify", {
> 
> 		address: contractImplementation,
> 
> 		constructorArguments: [],
> 
> 	})
> 
> ```

Big help, thanks friend

**Comment by tpae** [comment]
Why is this taking so long to merge

File: scripts/deploy.js (hunk 1 above)
Comment (MrMcGoats) [reply scripts/deploy.js:18]: Ya, that deploy script isn't for the upgradeable version. I forgot to include it. I'll add it, as well as the script to do the upgrade

**Comment by MrMcGoats** [comment]
The deploy-upgradeable script won't work as-is. The ERC721AUpgradeable contract is intended to be extended, so it isn't actually initializable  itself.

I can add a base contract that does nothing other than have in initilizer function, I guess, but not sure if that's something that would be useful.

File: scripts/deploy.js (hunk 1 above)
Comment (coindegen) [reply scripts/deploy.js:18]: > here's what I use, maybe it helps
> 	const contractImplementation = await getImplementationAddress(upgr.address)
> ```

thanks, very useful. I had no idea you could do this. For anyway else it comes from here:
```
import { getImplementationAddress } from '@openzeppelin/upgrades-core';
```

**Comment by coindegen** [comment]
> The deploy-upgradeable script won't work as-is. The ERC721AUpgradeable contract is intended to be extended, so it isn't actually initializable itself.
> 
> I can add a base contract that does nothing other than have in initilizer function, I guess, but not sure if that's something that would be useful.

I would find it useful. I made the mistake yesterday of deploying a ERC721A contract directly and was confused why I was missing a `mint` function. 😅

Quick question, does it generally cost more gas upfront to deploy the upgradeable version?

**Comment by MrMcGoats** [comment]
My assumption is that it does cost more gas, but I have not tested. I'll test it tomorrow and report back

**Comment by tpae** [comment]
What is the status of this? Need any help?

**Comment by MrMcGoats** [comment]
Just waiting to be merged at this point

**Comment by tpae** [comment]
@cygaar, could we get some feedback to help get this merged? we appreciate you 🙏

File: .nvmrc [hunk 2]
```diff
@@ -0,0 +1 @@
+v14.16.1
```
Comment (tpae) [inline .nvmrc:1]: current LTS is node 16. Could we remove this file?

File: contracts/mocks/ERC721AUpgradeableMock.sol [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: MIT
+// Created by: cygaar
+// Made upgradeable by: MrMcGoats
```
Comment (tpae) [inline contracts/mocks/ERC721AUpgradeableMock.sol:3]: appreciate the comments, but maybe we don't need this for the open source?

File: test/ERC721AUpgradeable.test.js [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+const GAS_MAGIC_VALUE = 20000;
+
+describe('ERC721A', function () {
+  beforeEach(async function () {
+    this.ERC721A = await ethers.getContractFactory('ERC721AUpgradeableMock');
+    this.ERC721Receiver = await ethers.getContractFactory('ERC721ReceiverMock');
+    this.erc721a = await this.ERC721A.deploy();
+    await this.erc721a.deployed();
+    // Initialize
+    this.erc721a.initialize('Azuki', 'AZUKI')
```
Comment (tpae) [inline test/ERC721AUpgradeable.test.js:15]: missing `;`

File: scripts/deploy.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  // Hardhat always runs the compile task when running scripts with its command
+  // line interface.
+  //
+  // If this script is run directly using `node` you may want to call compile
+  // manually to make sure everything is compiled
+  // await hre.run('compile');
+
+  // We get the contract to deploy
+  const Contract = await ethers.getContractFactory("ERC721A");
+  const contract = await Contract.deploy("Test", "TST", 100);
```
Comment (tpae) [inline scripts/deploy.js:18]: 3rd parameter is no longer needed

File: scripts/deploy-upgradeable.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  // await hre.run('compile');
+
+  // We get the contract to deploy
+  const [owner] = await ethers.getSigners();
+
+  const Contract = await ethers.getContractFactory("ERC721AUpgradeable");
+  const contract = await upgrades.deployProxy(Contract, /*[
+    "ERC721A",
+    "COIN",
+    10
```
Comment (tpae) [inline scripts/deploy-upgradeable.js:23]: 3rd parameter is no longer needed

File: package.json [hunk 7]
```diff
@@ -20,9 +20,14 @@
...
   "devDependencies": {
     "@nomiclabs/hardhat-ethers": "^2.0.4",
+    "@nomiclabs/hardhat-etherscan": "^3.0.0",
     "@nomiclabs/hardhat-waffle": "^2.0.1",
+    "@openzeppelin/contracts": "^4.4.2",
+    "@openzeppelin/contracts-upgradeable": "^4.4.2",
+    "@openzeppelin/hardhat-upgrades": "^1.13.0",
     "@openzeppelin/test-helpers": "^0.5.15",
     "chai": "^4.3.4",
+    "dotenv": "^12.0.4",
```
Comment (tpae) [inline package.json:30]: What's the purpose of the dotenv? could we remove this and keep this project agnostic to the user's dev environment?

File: .eslintrc.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+module.exports = {
```
Comment (tpae) [inline .eslintrc.js:1]: This file is not necessary, the repo has `.eslintrc` file already. Please remove

File: .npmignore [hunk 9]
```diff
@@ -0,0 +1,3 @@
+hardhat.config.ts
```
Comment (tpae) [inline .npmignore:1]: Let's remove this file

File: contracts/ERC721AUpgradeable.sol [hunk 10]
```diff
@@ -0,0 +1,508 @@
+// SPDX-License-Identifier: MIT
+// Creators: locationtba.eth, 2pmflow.eth
+// Made upgradeable by: mrmcgoats.eth for Xenum Technology (xenum.io)
```
Comment (tpae) [inline contracts/ERC721AUpgradeable.sol:3]: Let's keep the license and creator in-tact that follows the creator's guidelines. Appreciate your contribution, but not sure if we want that in the comments that are part of everyone's contracts moving forward. On the default `ERC721A.sol` file, it's simple like this:
```
// SPDX-License-Identifier: MIT
// Creator: Chiru Labs
```
Let's keep it that way

**Comment by Vectorized** [comment]
Thank you for your submission.

We have a transpiler based implementation in the works.

https://github.com/Vectorized/ERC721A-Transpiler
https://github.com/Vectorized/ERC721A-Upgradeable

We will credit you on Twitter for the next release.

**Pull request closed without merging**